source/agora/common/BlockStorage.d

```diff
@@ -159,35 +298,346 @@ public class BlockStorage
 
         Params:
             block = `Block` to read
+            height = height of `Block`
 
         Returns:
             Returns true if success, otherwise returns false.
 
     ***************************************************************************/
 
-    public bool readBlock (ref Block block, size_t height)
+    public bool readBlock (ref Block block, size_t height) @safe
     {
-        if (height >= this.block_positions.length)
+        if ((this.height_idx.length == 0) ||
+            (this.height_idx.back.height < height))
             return false;
 
-        const size_t next_height = height + 1;
-        this.map();
+        auto finds
+            = this.height_idx[].find!( (a, b) => a.height == b)(height);
+
+        if (!finds.empty)
+        {
+            const size_t position = finds.front.position;
 
-        const size_t from = this.block_positions[height];
-        size_t to;
+            this.map(position / MFILE_MAP_SIZE);
 
-        if (next_height >= this.block_positions.length)
-            to = file.length;
+            const size_t x0 = position % MFILE_MAP_SIZE;
+            const size_t block_size = this.readSizeT(MFILE_HEAD_SIZE + x0);
+
+            assert(x0 < this.data_size);
+
+            const size_t x2 = x0 + MFILE_HEAD_SIZE + size_t.sizeof;
+            const size_t x3 = x2 + block_size;
+
+            block = deserialize!Block(this.read(x2, x3));
+
+            return true;
+        }
         else
-            to = this.block_positions[next_height];
+        {
+            return false;
+        }
+    }
 
-        assert(from < file.length);
-        assert(from < to);
+    /***************************************************************************
 
-        block = deserialize!Block(cast(ubyte[])this.file[from .. to]);
+        Read block from the file.
 
-        return true;
+        Params:
+            block = `Block` to read
+            hash = `Hash` of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, Hash hash) @safe
```

Comment:
The whole point of memory mapping is to return a `const` pointer, so that there is only ever one copy of the block. This copy the block.